packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -3122,18 +3107,64 @@ export class StripeHelper extends StripeHelperBase {
 
     const productPaymentCycleOld = this.stripePlanToPaymentCycle(planOld);
 
+    const sameBillingCycle =
+      productPaymentCycleOld === baseDetails.productPaymentCycleNew;
+
+    // get next invoice details
+    const nextInvoice = upcomingInvoiceWithInvoiceItem
+      ? upcomingInvoiceWithInvoiceItem
+      : await this.previewInvoiceBySubscriptionId({
+          subscriptionId: subscription.id,
+        });
+
+    const { total: nextInvoiceTotal, currency: nextInvoiceCurrency } =
+      nextInvoice;
+
+    // https://stripe.com/docs/api/invoices/object#invoice_object-lines
+    // as per the Stripe docs, one of the individual line items includes
+    // new plan pricing and tax information without additional fees
+    const upcomingInvoiceLineItem =
+      nextInvoice.lines && nextInvoice.lines.data
+        ? nextInvoice.lines.data.find((line) => !line.proration)
+        : undefined;
+
+    // if same billing cycle and line item exists, add amount and tax for new total
+    // else return the next invoice total
+    const newTotalInCents =
+      sameBillingCycle &&
+      upcomingInvoiceLineItem &&
+      upcomingInvoiceLineItem.tax_amounts
+        ? upcomingInvoiceLineItem.amount +
+          upcomingInvoiceLineItem.tax_amounts[0].amount
+        : nextInvoiceTotal;
+
+    // calculated proration
+    // if same billing cycle, return next invoice total minus new plan price
+    // else return amount due on current invoice
+    const onetimePaymentAmount = sameBillingCycle
+      ? nextInvoiceTotal - newTotalInCents
+      : invoiceAmountDue;
+
+    // calculated old total
+    // if same billing cycle, get new invoiceTotal
+    // else get current invoiceTotal
+    const oldTotalInCents = sameBillingCycle
+      ? invoiceTotalNewInCents
```

Comment:
Is this assigning the new invoice total to the old total?

Reply:
I didn't rename the variable, but we can change it to something more descriptive if you want.